runtime/lua/vim/lsp/buf.lua

```diff
@@ -745,8 +745,20 @@ function M.code_action(options)
     context.triggerKind = vim.lsp.protocol.CodeActionTriggerKind.Invoked
   end
   if not context.diagnostics then
-    local bufnr = api.nvim_get_current_buf()
-    context.diagnostics = vim.lsp.diagnostic.get_line_diagnostics(bufnr)
+    context.diagnostics = {}
+    local diagnostics = vim.lsp.diagnostic._get(0)
+    local cursor = api.nvim_win_get_cursor(0)
+    local row, col = cursor[1] - 1, cursor[2]
```

Comment:
I think the `col` here needs to be converted considering the offset encoding of the client/server. If it's utf-16 or utf-32, the comparsion below could include or exclude the wrong diagnostics.

Somewhat related: It would probably be more efficient to fetch the diagnostic in vim-format, filter on them, and only afterwards convert to lsp diagnostic format, to do fewer conversions.